server/app/controllers/admin/AdminProgramStatusesController.java

```diff
@@ -34,6 +57,210 @@ public Result index(Http.Request request, long programId) throws ProgramNotFound
+            validateEditRequest(request, program);
+        editForm = validated.getLeft();
+        updatedStatus = validated.getRight();
+      } catch (MissingStatusException e) {
+        return redirect(routes.AdminProgramStatusesController.index(programId).url())
+            .flashing(
+                "error",
+                "The status being edited no longer exists and may have been modified in a separate"
+                    + " window.");
+      }
+      maybeEditForm = Optional.of(editForm);
+      if (!editForm.hasErrors() && updatedStatus.isPresent()) {
+        Result result = redirect(routes.AdminProgramStatusesController.index(programId).url());
+        ErrorAnd<ProgramDefinition, CiviFormError> setStatusResult =
+            service.setStatuses(programId, updatedStatus.get());
+        if (setStatusResult.isError()) {
+          String errorMessage = joinErrors(setStatusResult.getErrors());
+          return result.flashing("error", errorMessage);
+        }
+        return result.flashing(
+            "success",
+            previousStatusCount == updatedStatus.get().getStatuses().size()
+                ? "Status updated"
+                : "Status created");
+      }
+    }
+
+    return ok(statusesView.render(request, service.getProgramDefinition(programId), maybeEditForm));
+  }
+
+  private Pair<Form<ProgramStatusesEditForm>, Optional<StatusDefinitions>> validateEditRequest(
+      Http.Request request, ProgramDefinition program) throws MissingStatusException {
+    Form<ProgramStatusesEditForm> form =
+        formFactory
+            .form(ProgramStatusesEditForm.class)
+            .bindFromRequest(request, ProgramStatusesEditForm.FIELD_NAMES.toArray(new String[0]));
+    if (form.value().isEmpty()) {
+      return Pair.of(form, Optional.empty());
+    }
+    ProgramStatusesEditForm value = form.value().get();
+    StatusDefinitions current = program.statusDefinitions();
+    if (value.getOriginalStatusText().isEmpty()) {
+      try {
+        current.setStatuses(
+            addStatus(
+                current.getStatuses(),
+                StatusDefinitions.Status.builder()
+                    .setStatusText(value.getStatusText())
+                    .setLocalizedStatusText(
+                        LocalizedStrings.withDefaultValue(value.getStatusText()))
+                    .setEmailBodyText(value.getEmailBody())
+                    .setLocalizedEmailBodyText(
+                        Optional.of(LocalizedStrings.withDefaultValue(value.getEmailBody())))
+                    .build()));
+      } catch (DuplicateStatusException e) {
+        form = form.withError(ProgramStatusesEditForm.STATUS_TEXT_FORM_NAME, e.userFacingMessage);
+      }
+    } else {
+      try {
+        current.setStatuses(
+            replaceStatus(
+                current.getStatuses(),
+                value.getOriginalStatusText(),
+                (existingStatus) -> {
```

Comment:
This feels a bit convoluted as well this is the only call site so it's unclear why a function param is needed.

Passing in value, and having replaceStatus do this seems a lot cleaner, or passing in the 3 value data items explicitly which only adds 1 param.

replaceStatus(current.getStatuses(), value);